### PR TITLE
client: queue a delay cap flushing if there are ditry caps/snapcaps

### PR DESCRIPTION
**Pull request by lxbsz**
We should queue a cap release anyway if there are dirty caps/snapcaps for the inodes when trimming caps.

This together with https://github.com/ceph/ceph/commit/d4c175cd5e0f97ca75ee9e62592ac41e3da9c805 will fix:

 cluster:
    id:     8d2323ac-0820-49ad-b917-059ba00794f2
    health: HEALTH_WARN
            1 clients failing to respond to cache pressure

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -4784,6 +4784,9 @@ void Client::trim_caps(MetaSession *s, uint64_t max)
     // is deleted inside remove_cap
     ++p;
 
+    if (in->dirty_caps || in->cap_snaps.size())
+      cap_delay_requeue(in.get());
```
Comment (batrick) [inline src/client/Client.cc:4788]: I think also, if the Inode has to flush dirty caps then we should not include it it as "trimmed" (line 4827 below).

**Review by batrick — CHANGES_REQUESTED** [review]
Please create a tracker ticket too!

File: src/client/Client.cc (hunk 1 above)
Comment (vshankar) [reply src/client/Client.cc:4788]: Makes sense.

Comment (lxbsz) [reply src/client/Client.cc:4788]: Makes sense and I will fix it. Thanks @batrick @vshankar

File: src/client/Client.cc [hunk 2]
```diff
@@ -4833,7 +4836,8 @@ void Client::trim_caps(MetaSession *s, uint64_t max)
       }
       if (all && in->ino != CEPH_INO_ROOT) {
         ldout(cct, 20) << __func__ << " counting as trimmed: " << *in << dendl;
-	trimmed++;
```
Comment (vshankar) [inline src/client/Client.cc:4836]: Umm.. I thought the trimming for such inodes would be skipped altogether.

Comment (lxbsz) [reply src/client/Client.cc:4836]: I don't think so. Here we can just try to trim the dentries and then later after the dirty caps being flushed the clients should release the caps. This is also what kclient will do.

Comment (lxbsz) [reply src/client/Client.cc:4836]: I am thinking if we just skip trimming the `dentries` will be buggy ? Such as if after iterating all the inodes we still couldn't get enough inodes to release, and we just trigger to flush the dirty caps, then after the dirty caps triggered the clients couldn't release the caps by itself. And we must issue a new recall request to recall them.

Currently no matter there is any dirty caps we will always try to trim the dentries, and just after the dirty caps flushed the corresponding caps will be released.

Comment (vshankar) [reply src/client/Client.cc:4836]: > I don't think so. Here we can just try to trim the dentries and then later after the dirty caps being flushed the clients should release the caps. This is also what kclient will do.

You are correct. The dentry can be trimmed and that would help with the client dropping states when the mds requests a state recall. The inode itself gets deallocated at a later point when its ref drops to 0.

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#13-Oct-2023

**Comment by vshankar** [comment]
jenkins test windows

**Pull request merged**